### DISCUSSION TITLE
How can I test Interactive transactions?

### DISCUSSION THREAD

**Discussion by leoovogel** [prisma/prisma#14435] [Q&A]
Hello, I'm unit testing my application with JEST, but I ran into a difficulty when testing the interactive transactions.
In the example below, how would I test if prismaTransaction.account.update was called and how to test if an error was thrown if it satisfies the condition?

```ts
export async function accountWithdrawal(client: IClient, value: number) {
  const clientAccount = await prisma.account.findUniqueOrThrow({
    where: { client_id: client.id },
  });

  await prisma.$transaction(async (prismaTransaction) => {
    const result = await prismaTransaction.account.update({
      where: { client_id: client.id },
      data: {
        available_balance: {
          decrement: value,
        },
      },
    });

    if (Number(result.available_balance) < 0) {
      throw new HttpException('Insufficient funds', StatusCodes.BAD_REQUEST);
    }

    await prismaTransaction.transaction_history.create({
      data: {
        account_id: clientAccount.id,
        transaction_type: 'WITHDRAWAL',
        value,
      },
    });
  });

  return { message: 'Withdrawal successful' };
}
```

I've tried something close to this but to no avail:
```ts
  it('should throw an error if when making a withdrawal the balance is negative', async () => {
    jest.spyOn(prisma.account, "findUniqueOrThrow").mockResolvedValueOnce(mockAccount);
    jest.spyOn(prisma, "$transaction").mockResolvedValueOnce(
      jest.spyOn(prisma.transaction_history, "create").mockResolvedValueOnce({ available_balance: -5 })
    );

    const result = await accountWithdrawal(mockClient, 100);

    expect(result).not.toEqual({ message: 'Withdrawal successful' });
  });
  ```

**Comment by pantharshit00** [comment]
Hello @leonardoVogel , our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

If it is a single test, following might work depending on a usecase:
```ts
    jest.spyOn(prisma, '$transaction').mockImplementation((callback) => {
      return callback(prisma)
    })
```
and then mock methods on regular prisma objects. Caveats:

1. It won’t be running in actual transaction, so if anything goes wrong no changes will be rolled back. But I assume the goal is to avoid actual DB anyway, otherwise, mocking would’n have came up
2. Since we are passing actual prisma instance to the tx callback, it can do more than an actual transaction object will be able to: bunch of methods are not available on real tx proxies. TS will be able to prevent it, but there will be no corresponding runtime check

**Reply by vladi-stevanovic** [reply]
👋 Hello @leonardoVogel did you perhaps have a chance to check Nurul's recommendation? 

Let us know if you still have any questions, or feel free to "Mark as answer" if this discussion is now solved!
It would be super useful for other users facing this problem! 🙌

**Reply by gabrielgasp** [reply]
@pantharshit00 I'm facing the same problem and your solution kind of works but, as far as I understand, this simply mocks the behavior of the `$transaction` method, but doesn't assert that the database operations actually happened within the transaction callback.

Is there a way to address this?

Here is an example of what I mean:
```ts
export async function execute() {
  await prisma.$transaction(async (tx) => {
    console.log('empty transaction');
  });

  const result = await prisma.entity1.create({...});

  if (!result) {
    throw new Error()
  }

  await prisma.entity2.create({...});

  return true;
}
```

Edit: Found a workaround this as follows:

1. Write test cases for each database operation throwing an error (you are probably doing this already);
2. Check that the mocked `prisma.$transaction` rethrows the error thrown by the database operation. If this happens it means that said error was thrown inside the transaction callback.

Here is a simple example:
```ts
it('test', async () => {
  prisma.$transaction.mockImplementationOnce((cb) => cb(prisma));
  prisma.entity1.create.mockResolvedValueOnce('created');
  prisma.entity2.create.mockRejectedValueOnce('creation error');

  await yourFunction()

  await expect(prisma.$transaction.mock.results[0].value).rejects.toBe('creation error');
});
```

**Reply by david-almeida-capchase** [reply]
I'm actually facing a similar issue, where I want to introduce an error into a transaction and check that the rollback happened.

This is actually working for me doing this:

```
      jest
        .spyOn(prisma.someModel, 'createMany')
        .mockRejectedValue(new Error('nope'));
```
 Where the tx.someModel.createMany actually throws inside the transaction.

But this only works up to version 4.15, and I was eager to update to version 5.
@pantharshit00 do you know of a better mock that could fit my needs? Basically I'd like to throw an error in the end of the transaction to assert that the DB state is correct at the end (meaning, all changes were rolled-backed)

**Comment by Abiytsedeka** [comment]
it('test', async () => {
  prisma.$transaction.mockImplementationOnce((cb) => cb(prisma));
  prisma.entity1.create.mockResolvedValueOnce('created');
  prisma.entity2.create.mockRejectedValueOnce('creation error');

  await yourFunction()

  await expect(prisma.$transaction.mock.results[0].value).rejects.toBe('creation error');
});

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
